Initial IPv6 Support

taspelund opened the pull request:
This PR adds initial support for IPv6 static routing in maghemite, with the very beginnings of BGP support (really just originated IPv6 prefixes).
This addresses a few bugs found along the way (e.g. #529), adds unit tests for old/new functionality, and does a good bit of code cleanup, e.g.
 - mostly standardizing on oxnet::IpNet in generic spots and rdb::Prefix in db-specific spots, rather than several unique types
 - moving common functionality into mg-common where it can be used by all other crates w/o creating circular dependencies

taspelund commented on bgp/src/test.rs:
```diff
@@ -83,23 +42,28 @@ fn test_basic_peering() {
     // Ensure that r1's peer session to r2 has gone back to connect.
     r2.shutdown();
     d2.shutdown();
-    wait_for_eq!(r1_session.state(), FsmStateKind::Connect);
```
The main reason I modified this is because I believed the FSM state observed is timing-dependent, not consistent.

It seemed to me like the goal of this test case is to confirm that "shutdown" on one side brings the session down.
If so, then looking for "!established" is a less timing-dependent way to confirm that things have been completely torn down.

Otherwise, the state on the active side could be "connect", "opensent" (since the listener isn't torn down on the shutdown side, the transport can still come up) or "idle".
I was hoping to avoid a scenario where the test fails because the active peer is in "idle" or "opensent" and we were only looking for "connect".

I don't have super strong opinions about this though, so if reverting that change means the test passes consistently again, then I'm fine with that.

taspelund replied:
I'm not sure at what point this test became unreliable.  I'll go back through and figure out which of my changes borked things.  I don't suspect the issue is with the logic change of "state != established", but I could be wrong

rcgoodfellow commented on bgp/src/clock.rs:
```diff
@@ -2,77 +2,235 @@
+        lock!(self.value).remaining
+    }
+
+    /// Reset the value of a timer to the timer's interval.
```
I'm not sure this is what I would call jitter, as to me jitter means the deviation from the target interval varies from cycle to cycle. This is more like a phase offset? I guess if `reset` is called often, then this is more jitter like.

rcgoodfellow commented on bgp/src/connection.rs:
```diff
@@ -21,63 +26,189 @@ pub const MAX_MD5SIG_KEYLEN: usize = 80;
+    }
+}
+
+impl PartialOrd for ConnectionId {
```
Just use`#[derive(PartialOrd)]`?

rcgoodfellow commented on bgp/src/connection.rs:
```diff
@@ -21,63 +26,189 @@ pub const MAX_MD5SIG_KEYLEN: usize = 80;
 
-    /// Try to connect to a peer. On success messages from the peer a
-    /// propagated through `event_tx`.
+/// Implementors of this trait initiate outbound BGP connections to peers.
```
This comment seems like it's more in reference to an implementation of this trait rather than the trait itself? Similar for the comment on `connect` below.

rcgoodfellow commented on bgp/src/connection_tcp.rs:
```diff
@@ -97,348 +93,675 @@ impl BgpListener<BgpConnectionTcp> for BgpListenerTcp {
-        let (conn, mut peer) = self.listener.accept()?;
+        let (conn, mut peer) = self.listener.accept().map_err(|e| {
+            if e.kind() == std::io::ErrorKind::WouldBlock {
+                Error::Timeout
```
Going immediately to a timeout error on `WouldBlock` does not seem right. Shouldn't we be retrying until for a `timeout` duration?

rcgoodfellow commented on bgp/src/connection_tcp.rs:
```diff
@@ -97,348 +93,675 @@ impl BgpListener<BgpConnectionTcp> for BgpListenerTcp {
-                lock!(self.conn).replace(new_conn.try_clone()?);
-                if let Some(ttl) = min_ttl {
-                    self.set_min_ttl(ttl)?;
+            // Start SA tracking and keepalive for Illumos MD5
```
Why are we separating this from initialization and creating this mostly duplicate logic? Initialization and kicking off the keepalive thread happened all in one spot.

rcgoodfellow commented on bgp/src/connection_tcp.rs:
```diff
@@ -97,348 +93,675 @@ impl BgpListener<BgpConnectionTcp> for BgpListenerTcp {
+            let conn_id = self.id;
+
+            // Clone the TcpStream for the recv thread
+            if let Ok(conn) = lock!(self.conn).try_clone() {
```
Handle the failure case?

rcgoodfellow commented on bgp/src/connection_tcp.rs:
```diff
@@ -97,348 +93,675 @@ impl BgpListener<BgpConnectionTcp> for BgpListenerTcp {
+            return;
+        }
+
+        // Take the params (they should exist since we haven't started yet)
```
I think using the typestate pattern would help tighten this up. If for some reason this expectation fails, there is no logging here and we'll have no idea.

rcgoodfellow commented on bgp/src/connection_tcp.rs:
```diff
@@ -547,10 +891,18 @@ impl BgpConnectionTcp {
         msg: Message,
     ) -> Result<(), Error> {
-        trace!(log, "sending {:#?}", msg);
+        // connection_log_lite!(log,
```
remove commented out code or un-comment out?

rcgoodfellow commented on bgp/src/connection_tcp.rs:
```diff
@@ -666,34 +1020,47 @@ impl BgpConnectionTcp {
-        debug!(self.log, "spawning security association keepalive loop");
+        connection_log!(self,
+            debug,
+            // space after loop is needed... because macros?
```
what was the issue here?

rcgoodfellow commented on bgp/src/connection_tcp.rs:
```diff
@@ -829,3 +1206,86 @@ fn sa_set(src: SocketAddr, dst: SocketAddr) -> [(SocketAddr, SocketAddr); 4] {
+    // Set MD5 security associations for Illumos
+    for local in locals.iter() {
+        for (a, b) in sa_set(*local, peer) {
+            if let Err(e) = libnet::pf_key::tcp_md5_key_add(
```
Is it safe to assume that this will always be an add? What if the connection rolled over and we are back here inside the keepalive time? In that case we would need to handle the update case.

taspelund replied:
From a functional standpoint, it appears that always calling `add` works just fine.
After leaving the interop topology running for 4+ days, the kernel only returns 8 SAs when checking the output of `tcpkey {dump|save}`, which is correct given we expect the SA configuration to setup 4 per peer.
e.g.
For a connection between us (169.254.20.1) and a peer (169.254.20.2), I see these SAs:
1. src=169.254.20.1, sport=0, dst=169.254.20.2, dport=179
2. src=169.254.20.1, sport 179, dst=169.254.20.2, dport=0
3. src=169.254.20.2, sport=0, dst=169.254.20.1, dport=179
4. src=169.254.20.2, sport 179, dst=169.254.20.1, dport=0

If there was a leak, I would suspect there to be FAR more than 8 SAs given we use a lifetime of 2 minutes.

rcgoodfellow commented on bgp/src/log.rs:
```diff
@@ -0,0 +1,346 @@
+        slog::$level!($self.log,
+            $msg;
+            "component" => crate::COMPONENT_BGP,
+            "module" => crate::MOD_NEIGHBOR,
```
Every log here has both

```rust
"component" => crate::COMPONENT_BGP
```

and

```rust
"module" => crate::MOD_NEIGHBOR
```

so what is the latter telling us?

rcgoodfellow commented on bgp/src/messages.rs:
```diff
@@ -425,6 +506,20 @@ impl OpenMessage {
 
+impl Display for OpenMessage {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        let mut param_string = String::new();
```
this can probably be streamlined to

```rust
self.parameters
  .iter()
  .map(ToString::to_string)
  .collect::<Vec<_>>()
  .join(",")
```

rcgoodfellow commented on bgp/src/messages.rs:
```diff
@@ -2575,11 +2911,179 @@ pub enum Safi {
+                }
+            }
+            Err(_) => {
+                // Fallback for wire conversion errors
```
Maybe this should be `TryFrom` instead. What are the potential ramifications if we fail to convert between versions and default to 0.0.0.0/0 below?

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -203,185 +475,214 @@ pub enum FsmEvent<Cnx: BgpConnection> {
+            allow_export: ImportExportPolicy::default(),
+            vlan_id: None,
+            // Default timer values based on BGP standards
+            connect_retry_time: Duration::from_secs(120),
```
This feels a bit dangerous. I guess keepalive and hold will get negotiated, but implicit defaults for the others still concern me. The only place I see `default() being used in `bgp_admin` these values are available in the `Neighbor` request object. I'm also wondering why timers are now a part of the session info object.

rcgoodfellow commented on bgp/src/router.rs:
```diff
@@ -170,14 +193,33 @@ impl<Cnx: BgpConnection + 'static> Router<Cnx> {
         info: SessionInfo,
     ) -> Result<Arc<SessionRunner<Cnx>>, Error> {
-        a2s.insert(peer.host.ip(), event_tx.clone());
+        // Update the SessionInfo with timer values from peer config
```
It seems strange to do this here. If we're going to merge this information into `SessionInfo` can we just construct `SessionInfo` with complete data from the outset?

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -569,19 +1082,180 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+            // Lock is released here at end of scope
+        }
+
+        // Spawn new connector with NO locks held (critical for avoiding deadlock)
```
What deadlock is being avoided? I see `self.connector_handle` being held above, but that's not available to `Cnx::Connector::connect`. `self.connector_handle` is locked again at 1178, but it seems like it could just be held the whole time? Am I missing something?

taspelund replied:
I don't recall the exact flow where this was relevant, as it was something I found and fixed during some fairly quick code iteration.
I want to say the deadlock was happening when passing `lock!(self.session)` as an argument to Connector::connect(), but I ended up restructuring it so that wouldn't be an issue.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -569,19 +1082,180 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+        *lock!(self.primary) = None;
+    }
+
+    /// This is the BGP peer state machine entry point. This function only
```
@rcgoodfellow you are here.

rcgoodfellow reviewed (commented):
Thanks for all the work in putting this together @taspelund! I'm through BGP up to the FSM. Will pick back up there tomorrow.

taspelund commented on bgp/src/connection.rs:
```diff
@@ -21,63 +26,189 @@ pub const MAX_MD5SIG_KEYLEN: usize = 80;
+        }
+    }
+
+    pub fn direction(&self) -> &'static str {
```
I thought about this recently too.  I would honestly rather just rip out ConnectionCreator and replace it with ConnectionDirection, as I think that has more meaning for anyone who isn't already familiar with the codebase.

taspelund commented on bgp/src/connection_tcp.rs:
```diff
@@ -97,348 +93,675 @@ impl BgpListener<BgpConnectionTcp> for BgpListenerTcp {
-                        self.log,
-                        "source address selection for {}: {e}", self.peer
+                    connection_log_lite!(log,
+                        debug,
```
I would argue it should be `notice` rather than warning, since this would be normal and require no action from the operator if the peer was simply shutdown.

Slog doesn't have a macro for notice, so we either do info or warn. I'd lean towards info, personally, but I don't have strong feelings about it.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -569,19 +1084,188 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+            let mut handle_guard = lock!(self.connector_handle);
+
+            if let Some(old_handle) = handle_guard.take() {
+                // Check if thread is still running
```
Since the connect timeout is 1/3 of the connect retry timer, it seems we should not be able to get here through a connect retry. Is this handling a case where there is a failure that short-circuits us here in a lap through idle where idle hold time is zero. E.g. what does the path look like to land us here with a connect thread still running.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -1915,46 +6289,244 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                if let Some(c2) = conn2 {
+                    self.send_admin_reset_notification(c2);
+                }
+                self.counters
```
Can you explain the logic behind the counter updates for the various stop reasons?

taspelund replied:
Yeah, and I should probably add in a block comment that explains it too.

The idea here is to have a single helper method to clean up connections that need to be stopped.
When looking at the FSM description in the RFC, there are a lot of commonalities as to what items need to be updated upon a given error condition, regardless of what FSM state that error condition was encountered in.
This method is intended to centralize much of these situations into a single method which uses the `StopReason` enum to indicate which set of items need to be updated.

I also found that pretty much anytime stop() was being called, a connection was being unregistered as well.  So I added the connection registry deletion to this method as well; again, as a way to centralize that handling.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                                "{} event not allowed in this state, fsm transition to idle",
+                                session_event.title()
+                            );
+                            return FsmState::Idle;
```
Do we really want to transition to idle here - or try to correct the situation by just stopping the idle hold timer and carrying on in connect?

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                    session_log_lite!(self, error, "event rx error: {e}";
+                        "error" => format!("{e}")
+                    );
+                    // TODO: Possible death loop. Should we just panic here?
```
So the only other error condition here is `RecvTimeoutError::Disconnected` which means that [all of the senders have been disconnected/dropped](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=ca1dfe68943344d5445ee0f797802894).  That does mean that this will be a death loop with `continue` if we hit that error.

The channel is created all the way up in `bgp_admin::add_neighbor`, there the tx side is cloned and delved out. Following the setup path and all the tx clones that are made, one of the clones is held as a member of the `SessionRunner` itself, so it really should be impossible for this to happen.

I would probably put an `unreachable!()` statement here with a comment that explains why this should be an unreachable path.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                                error,
+                                new,
+                                conn,
+                                "rejected new {new_direction} connection ({}): multiple {new_direction} connections not allowed",
```
very long line. per rustfmt.toml we're trying to keep things within 80 cols. it seems rustfmt does not catch this for macros. There are quite a few other macro lines beyond 80 cols.

taspelund replied:
I agree with 80 cols, but in the case of string literals I think an exception should be made.
It's very difficult to grep for log strings if they are split/escaped across multiple lines.

taspelund replied:
I was referring to using grep on the source, not a log file.  Making the string multi-line via escaped newlines makes it harder to grep for the logging call.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                        ConnectionEvent::Message { msg, ref conn_id } => {
+                            match self.get_conn(conn_id) {
+                                Some(connection) => {
+                                    if connection.id() != conn.id() {
```
The only place I see connections being added to `self.connections` is in `register_conn` where connections are inserted with a key that is equal to their connection `id()`. When would we encounter a connection entry where the key is not equal to the `id()` of the value?

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                                        continue;
+                                    }
+                                }
+                                None => {
```
Can you add a comment on what gets us here?

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                    "event rx error: {e}";
+                    "error" => format!("{e}")
+                );
+                //TODO possible death loop. Should we just panic here?
```
[Ditto](https://github.com/oxidecomputer/maghemite/pull/533/files#r2483088217)

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                    info,
+                    new,
+                    exist.conn,
+                    "collision detected: new connection [{:?}, conn_id: {}], existing connection [{:?}, conn_id: {}]",
```
Should this log and the next indicate whether we coming from OpenConfirm versus OpenSent?

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                            Some(&exist.conn),
+                            StopReason::FsmError,
+                        );
                         return FsmState::Idle;
```
[Ditto](https://github.com/oxidecomputer/maghemite/pull/533/files#r2483047453).

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                            Some(&exist.conn),
+                            StopReason::FsmError,
+                        );
                         return FsmState::Idle;
```
[Ditto](https://github.com/oxidecomputer/maghemite/pull/533/files#r2483047453).

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+        }
+    }
+
+    /// Handler for collisions when existing connection was in OpenSent state
```
@rcgoodfellow you are here

rcgoodfellow reviewed (commented):
Next round of comments follows.

taspelund commented on bgp/src/connection_tcp.rs:
```diff
@@ -596,37 +956,31 @@ impl BgpConnectionTcp {
         locals: Vec<SocketAddr>,
         peer: SocketAddr,
     ) -> Result<(), Error> {
+        // First set up the socket using the standalone function
```
PR opened to include md5 password in interop here: https://github.com/oxidecomputer/testbed/pull/208

I do think we want more thorough testing for this than just confirming that interop comes up and establishes (like you mentioned in a different comment, making sure key rotation works on illumos is a key (no pun intended) item to ensure coverage for)

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+        }
+    }
+
+    /// Collision Resolution logic.
```
This comment does not really line up with what's going on in this function. This function is associating the connection id for an incoming message with either a new or existing connection in a collision context.

taspelund replied:
I'm sure I missed updating some comments when I refactored collision resolution, but I'm not sure how this comment ended up on this method at all.
Thanks for spotting it, I'll clean it up

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                                            return FsmState::OpenSent(new);
+                                        }
+
+                                        // Resolve collision on first Open - don't wait for both
```
IIUC this means that if new wins, we ditch the successful open for a connection that may later have a failed open?

rcgoodfellow replied:
My observation here is that if we view the established connection and the new connection as both working toward getting to established, in this case we may be preferring a connection that is less far along over one that has gotten further. And it seems odd to resolve a collision between two connections that have not reached equal footing in terms of the their position in the state machine yet.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                        // Event 12
+                        | ConnectionEvent::DelayOpenTimerExpires(
+                            conn_id,
+                        ) => {
```
Similar to the comments on dropping to idle on unexpected timer events. This feels like a bug. Should we be amplifying the impact of the bug by stopping connections in response and letting the bug influence the collision result like this?

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -1290,249 +4912,908 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                        session_event.title()
+                    );
+                    self.stop(Some(&pc.conn), None, StopReason::FsmError);
+                    self.exit_established(pc)
```
[Ditto](https://github.com/oxidecomputer/maghemite/pull/533/files#r2483047453).

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -1290,249 +4912,908 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                        "rx delay open timer expires, fsm transition to idle"
+                    );
+                    self.stop(Some(&pc.conn), None, StopReason::FsmError);
+                    self.exit_established(pc)
```
[Ditto](https://github.com/oxidecomputer/maghemite/pull/533/files#r2483047453).

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -1290,249 +4912,908 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                            .get_conn(conn_id)
+                            .filter(|c| conn_timer!(c, keepalive).enabled())
+                        {
+                            self.stop(Some(&conn), None, StopReason::FsmError);
```
Stopping the connection feels unrelated to the fact that its keepalive timer expired? The only timer that is stopped in `stop` for `FsmError` is the connect retry timer.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -1290,249 +4912,908 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                            "rx delay open timer expires, fsm transition to idle"
+                        );
+                        self.stop(Some(&pc.conn), None, StopReason::FsmError);
                         self.exit_established(pc)
```
[Ditto](https://github.com/oxidecomputer/maghemite/pull/533/files#r2483047453).

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -1290,249 +4912,908 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                            .get_conn(conn_id)
+                            .filter(|c| conn_timer!(c, delay_open).enabled())
+                        {
+                            self.stop(Some(&conn), None, StopReason::FsmError);
```
Stopping the connection feels unrelated to the fact that its delay open timer expired? The only timer that is stopped in `stop` for `FsmError` is the connect retry timer.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -1290,249 +4912,908 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                                None,
+                                StopReason::FsmError,
+                            );
+                            self.exit_established(pc)
```
Previously we just logged this as a warning and carried on. I can see this happening if we got an active connection quickly and an open message got sent from a peer's active FSM path before it has a chance to halt it's own active FSM path.

taspelund replied:
One of the major motivations for the ConnectionId design was to avoid issues like that, which is accomplished by a couple different ideas working together:
1. Every BgpConnection has a unique ID, and all events specific to that connection include the ID as part of the connection
2. The FSM validates the event's ConnectionId before honoring the event.
3. The recv loop for a BgpConnection is started _by the FSM_ after passing semantic validation (e.g. are we in an active collision? are we already established?). This ensures no Messages are admitted into the event queue unless the FSM accepts the connection that message is tied to.

This specific codepath is handling an Open received in Established that isn't tied to a ConnectionId owned by this FSM state (pc.conn) or at least known in the registry (e.g. a connection surviving ConnectionCollision state per the long-winded comment above).
Any time we receive an Open that isn't one of the above circumstances is a legitimate FSM error perpetrated by the peer, and I think we should treat it as such.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -1290,249 +4912,908 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                                    self,
+                                    info,
+                                    pc.conn,
+                                    "rx collision resolution notification (conn_id: {}), fsm transition to active",
```
We are actually transitioning to Idle below.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -1698,34 +6055,23 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
-                    .lock()
-                    .unwrap()
-                    .as_secs() as u16,
+                conn_timer!(conn, hold).interval.as_secs() as u16,
```
For an open message, shouldn't this be the configured interval and not the negotiated one? Same for the two octet ASN case below.

taspelund replied:
If we're sending an Open, we have not performed negotiation yet.  This is true so long as DelayOpen is not implemented.
Currently we do the negotiation in handle_open(), which could probably be refactored into its own method so it's not strictly tied to receipt of an Open.

I definitely think it would be better to pull this from config rather than relying on the implementation details of "we haven't negotiated yet," so I'll push a change for this.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -1915,46 +6289,244 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+        match msg {
+            MessageKind::Open => {
+                self.counters.opens_received.fetch_add(1, Ordering::Relaxed);
+                self.counters
```
All these unexpected counter bumps should only be in the `if unexpected` branch below?

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -2135,20 +6738,24 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
 
     fn apply_static_update_policy(&self, update: &mut UpdateMessage) {
-        if self.is_ebgp() {
+        if let Some(ebgp) = self.is_ebgp()
```
Consider `self.is_ebgp().unwrap_or(false)`.

rcgoodfellow reviewed (commented):
Comments for the rest of session.rs.

rcgoodfellow commented on mg-api/src/lib.rs:
```diff
@@ -168,28 +172,74 @@ pub trait MgAdminApi {
         request: Query<AsnSelector>,
     ) -> Result<HttpResponseDeleted, HttpError>;
 
+    #[endpoint { method = PUT, path = "/bgp/config/origin6", versions = VERSION_IPV6_BASIC.. }]
```
Note that we only really need to version bump API endpoints that are directly consumed by omicron. Endpoints whose only consumer is `mgadm` don't need multiple versions.

rcgoodfellow commented on rdb/src/db.rs:
```diff
@@ -125,38 +149,184 @@ impl Db {
+        }
+    }
+
+    #[allow(dead_code)]
```
Was it intentional to leave these unused functions here?

taspelund replied:
Doesn't look like they're worth keeping -- the static_rib variants above either.  Going to remove them.

Thanks for pointing them out!

rcgoodfellow commented on rdb/src/db.rs:
```diff
@@ -464,48 +842,84 @@ impl Db {
                 };
 
                 let key = String::from_utf8_lossy(&key);
+                // XXX: figure out how to handle removal of old static routes
```
See my earlier comment on non-zero host bits.

rcgoodfellow commented on rdb/src/types.rs:
```diff
@@ -519,3 +697,179 @@ impl From<Prefix6> for PrefixChangeNotification {
+}
+
+/// Trait for encoding/decoding prefixes to/from BGP wire format
+pub trait BgpWireFormat: Sized {
```
Feels like this should be in the `bgp` crate? Maybe in `bgp::messages`?

rcgoodfellow commented on rdb/src/types.rs:
```diff
@@ -519,3 +697,179 @@ impl From<Prefix6> for PrefixChangeNotification {
+        Ok(buf)
+    }
+
+    fn from_wire(input: &[u8]) -> Result<(&[u8], Self), Self::Error> {
```
If this were in `bgp::messages` you could probably reuse some of the nom parser combinators for this.

rcgoodfellow reviewed (commented):
Thanks @taspelund! This set of comments concludes my initial review.

This is a ton of work. Thanks for bringing this all together. The BGP FSM bits in particular are a significant improvement, both in the collision handling and the overall expression of the state machine.

Looking forward to getting this across the line.

taspelund commented on bgp/src/session.rs:
```diff
@@ -177,16 +232,233 @@ pub enum FsmEvent<Cnx: BgpConnection> {
+    DelayOpenTimerExpires(ConnectionId),
+}
+
+impl fmt::Debug for ConnectionEvent {
```
Good point. I've moved to deriving debug for all enum *Event types

taspelund commented on bgp/src/session.rs:
```diff
@@ -177,16 +232,233 @@ pub enum FsmEvent<Cnx: BgpConnection> {
+    /// peer and the local system has sent an ack.
+    /// i.e.
+    /// We have ACKed the peer's connection.
+    /// We use this event to indicate an inbound connection has completed.
```
I think that comment may have been re-used from its original definition ;-)

I'll fix it

taspelund commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+
+                FsmEvent::Connection(connection_event) => {
+                    match connection_event {
+                        // XXX: Make sure we always log the message type + conn_id
```
This was a note to self.  Will confirm it's been addressed and will yank it out.

taspelund commented on bgp/src/session.rs:
```diff
@@ -1855,7 +6215,9 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
         }
 
-        if self.is_ibgp() {
+        if let Some(ibgp) = self.is_ibgp()
```
The reason I have that method setup to return `Option<bool>` is in case the caller makes a query while the peer is down (or before we know if they're internal/external).
Maybe that doesn't make the most sense ergonomically, but it felt like the idea of answering "is this session iBGP?" with `false` when the session is down seemed semantically wrong to me.

What are your thoughts on how to approach that?

Also, do you think the is_ibgp()/is_ebgp() methods should account for the optional configured ASN? `self.session.remote_asn`

taspelund commented on rdb/src/types.rs:
```diff
@@ -200,6 +215,12 @@ impl Ord for Prefix4 {
 impl Prefix4 {
+    pub fn new(ip: Ipv4Addr, length: u8) -> Self {
+        let mut new = Self { value: ip, length };
+        new.unset_host_bits();
```
Do you mean a docstring? or an update to public docs?

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
-                        self.bind_addr,
-                        self.neighbor.host,
-                        self.log.clone(),
+            let event = match self.event_rx.recv_timeout(IO_TIMEOUT) {
```
Something that encapsulates the next event extraction logic and makes that code reusable.

rcgoodfellow commented on bgp/src/session.rs:
```diff
@@ -686,568 +1393,3477 @@ impl<Cnx: BgpConnection + 'static> SessionRunner<Cnx> {
+                         *
+                         *   - changes its state to OpenSent.
+                         *
+                         * A HoldTimer value of 4 minutes is suggested.
```
My comment is about this specific 4 minute guidance that should be removed. It's not generally valid guidance.

taspelund commented:
Peers resetting the session in a loop: ✅ 
Local system resetting sessions in a loop: ✅ 
Route refresh (tested bidirectionally): ✅ 
GR: upon further inspection of our code, GR helper has not been implemented.  Routes are only marked stale when we send a Route Refresh (with the assumption that either the peer re-advertises us routes and un-marks them stale, or they get removed by the db reaper)
MD5 mismatch: ✅ 
Changing MD5 password: ✅ 
Toggling MD5 password: ✅ 

The manual testing of resetting sessions in a loop did a few thread leaks (issues in the shutdown flag logic) which I've fixed. I've also updated the MD5 SA Keepalive thread logic to have shorter sleeps so we aren't stuck waiting 60 seconds for the shutdown flag to get noticed (manual testing showed that flapping the session ~1/s resulted in ~60 threads lingering per peer, which makes sense given the old sleep was 60s).

With that all done, I would say everything is looking good and is ready for final review.
